airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/BufferManager.java

```diff
@@ -108,40 +113,74 @@ public Map<StreamDescriptor, MemoryBoundedLinkedBlockingQueue<AirbyteMessage>> g
       return new HashMap<>(buffers);
     }
 
-    public MemoryBoundedLinkedBlockingQueue<AirbyteMessage> getBuffer(final StreamDescriptor streamDescriptor) {
+    private MemoryBoundedLinkedBlockingQueue<AirbyteMessage> getBuffer(final StreamDescriptor streamDescriptor) {
       return buffers.get(streamDescriptor);
     }
 
-    public long getTotalGlobalQueueSizeInMb() {
+    public long getTotalGlobalQueueSizeBytes() {
       return buffers.values().stream().map(MemoryBoundedLinkedBlockingQueue::getCurrentMemoryUsage).mapToLong(Long::longValue).sum();
     }
 
-    public long getQueueSizeInMb(final StreamDescriptor streamDescriptor) {
+    public long getQueueSizeInRecords(final StreamDescriptor streamDescriptor) {
+      return getBuffer(streamDescriptor).size();
+    }
+
+    public long getQueueSizeBytes(final StreamDescriptor streamDescriptor) {
       return getBuffer(streamDescriptor).getCurrentMemoryUsage();
     }
 
     public Optional<Instant> getTimeOfLastRecord(final StreamDescriptor streamDescriptor) {
       return getBuffer(streamDescriptor).getTimeOfLastMessage();
     }
 
-    public Batch take(final StreamDescriptor streamDescriptor, long bytesToRead) {
-      var queue = buffers.get(streamDescriptor);
-      return new Batch(List.of());
-    }
+    public Batch take(final StreamDescriptor streamDescriptor, final long bytesToRead) {
+      final var queue = buffers.get(streamDescriptor);
 
-    class Batch implements AutoCloseable {
+      final AtomicLong bytesRead = new AtomicLong();
+        final var s = Stream.generate(() -> {
+          try {
+            return queue.poll(5, TimeUnit.MILLISECONDS);
+          } catch (final InterruptedException e) {
+            throw new RuntimeException(e);
+          }
+        }).takeWhile(memoryItem -> {
```

Comment:
I love and hate this.

Reply:
lol. we don't have to do streams if you don't want. this was the easiest version i could figure.

Reply:
no it's ok!

it just easy to misread this - requires knowing `takeWhile` is a terminal function. more of my complain with the streams api than anything else.